Fix assertion in ScanLowering for num_ctas>1

safelix opened the pull request:
Enable within-CTA scans if `num_ctas>1` and `cluster_dims[axis]==1` on Hopper or later.

Assume the `clusterCTAId` along the scan axis is `cctaIdAxis==0`, raise runtime assertion otherwise. Combine the `clusterCTAId` across the scan axis (`cctaIdParallel`) into the `flatIdParallel` and compute `numParallelLane` per CGA instead of per CTA. 


This fixes assertions when `num_ctas > 1`:
https://github.com/triton-lang/triton/blob/cea35daf3578767f45db9904f8437ed96d2dfaa8/lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp#L150 https://github.com/triton-lang/triton/blob/cea35daf3578767f45db9904f8437ed96d2dfaa8/lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp#L254



<!---
The core Triton is a small number of people, and we receive many PRs (thank
you!).  To help us review your code more quickly, **if you are a new
contributor (less than 3 PRs merged) we ask that you complete the following
tasks and include the filled-out checklist in your PR description.**

Complete the following tasks before sending your PR, and replace `[ ]` with
`[x]` to indicate you have done them.
-->

# New contributor declaration
- [x] I am not making a trivial change, such as fixing a typo in a comment.

- [x] I have written a PR description following these
  [rules](https://cbea.ms/git-commit/#why-not-how).

- [x] I have run `pre-commit run --from-ref origin/main --to-ref HEAD`.

- Select one of the following.
  - [x] I have added tests.
    - `/test` for `lit` tests
    - `/unittest` for C++ tests
    - `/python/test` for end-to-end tests
  - [ ] This PR does not need a test because see above.

- Select one of the following.
  - [x] I have not added any `lit` tests.
  - [ ] The `lit` tests I have added follow these [best practices](https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices),
    including the "tests should be minimal" section. (Usually running Python code
    and using the instructions it generates is not minimal.)

safelix commented:
Added the test cases, they pass if the scan is within a CTA and fail if it is across multiple CTAs.

I didn't find any logic which performs accumulation across CTAs in the code. I could look into the logic and test coverage for cross-CTA scans in a different PR tomorrow, but I think fixing the assertion for within-CTA scans can be independent from that.

How should I proceed?

Jokeren commented:
> Added the test cases, they pass if the scan is within a CTA and fail if it is across multiple CTAs.

For now it should happen only within a CTA. Please do not work on anything across CTAs

safelix commented:
The initial fix was only superficial, so I implemented within-CTA scan from scratch. I changed testing from end-to-end tests to layout tests to make sure all edge cases are caught. Pytest output:
```400 passed, 480 skipped, 19203 deselected in 147.85s (0:02:27)```

I'm not sure if the tests are too extensive now, maybe we could just test the combinations with `thread_size==1`?

Is there anything else I can do to get this merged?

ThomasRaoux commented on lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp:
```diff
@@ -488,8 +498,19 @@ ScanOpConversion::emitFastScan(triton::ScanOp op, triton::ScanOpAdaptor adaptor,
+  auto [laneIdAxis, warpIdAxis, flatIdParallel, cctaIdAxis, cctaIdParallel] =
+      getDelinearizedIds(rewriter, helper, laneId, warpId, cctaId);
+
+  // We assume that cctaIdAxis==0 and gather cctaIdParallel into the
+  // flatIdParallel. This requires numParallelLane to be computed per
+  // CGA instead of per CTA in storeWarpAccumulator() and AddPartialReduce().
+  rewriter.create<triton::AssertOp>(
+      loc, b.icmp_eq(cctaIdAxis, b.i32_val(0)),
+      StringRef("(cctaIdAxis==0) && Scan axis accross CTAs is not supported."));
```
we definitely don't want to create asserts as those can have a high cost

ThomasRaoux commented on lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp:
```diff
@@ -488,8 +498,19 @@ ScanOpConversion::emitFastScan(triton::ScanOp op, triton::ScanOpAdaptor adaptor,
 
-  auto [laneIdAxis, warpIdAxis, flatIdParallel] =
-      getDelinearizedIds(rewriter, helper, laneId, warpId);
+  auto [laneIdAxis, warpIdAxis, flatIdParallel, cctaIdAxis, cctaIdParallel] =
+      getDelinearizedIds(rewriter, helper, laneId, warpId, cctaId);
```
I don't think the math should consider `cctaId` if the scan is not happening cross ctas

ThomasRaoux commented on python/test/unit/language/test_core.py:
```diff
@@ -2699,7 +2700,55 @@ def kernel(X, Y, N, BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr):
-    BlockedLayout([1, 2], [1, THREADS_PER_WARP // 1], [1, 4], [1, 0], [1, 1], [1, 1], [0, 1]),
+    BlockedLayout([1, 4], [1, THREADS_PER_WARP // 1], [1, 4], [1, 0], [1, 1], [1, 1], [0, 1]),
+    # thread_size=4, num_warps=1, num_ctas=4: CTASplitNum=[1,1]
+    BlockedLayout([1, 4], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [4, 1], [1, 1], [0, 1]),
+    BlockedLayout([1, 4], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [4, 1], [1, 1], [0, 1]),
+    BlockedLayout([4, 1], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [1, 4], [1, 1], [0, 1]),
+    BlockedLayout([2, 2], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [2, 2], [1, 1], [0, 1]),
+    BlockedLayout([2, 2], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [2, 2], [1, 1], [0, 1]),
+    BlockedLayout([1, 4], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [4, 1], [1, 1], [1, 0]),
+    BlockedLayout([1, 4], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [4, 1], [1, 1], [1, 0]),
+    BlockedLayout([4, 1], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [1, 4], [1, 1], [1, 0]),
+    BlockedLayout([2, 2], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [2, 2], [1, 1], [1, 0]),
+    BlockedLayout([2, 2], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [2, 2], [1, 1], [1, 0]),
+    BlockedLayout([1, 4], [1, THREADS_PER_WARP // 1], [1, 1], [0, 1], [1, 4], [1, 1], [1, 0]),
+    # thread_size=4, num_warps=1, num_ctas=4: CTASplitNum=CTAsPerCGA
+    BlockedLayout([1, 4], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [4, 1], [4, 1], [0, 1]),
+    BlockedLayout([1, 4], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [4, 1], [4, 1], [0, 1]),
+    BlockedLayout([4, 1], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [1, 4], [1, 4], [0, 1]),
+    BlockedLayout([2, 2], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [2, 2], [2, 2], [0, 1]),
+    BlockedLayout([2, 2], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [2, 2], [2, 2], [0, 1]),
+    BlockedLayout([1, 4], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [4, 1], [4, 1], [1, 0]),
+    BlockedLayout([1, 4], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [4, 1], [4, 1], [1, 0]),
+    BlockedLayout([4, 1], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [1, 4], [1, 4], [1, 0]),
+    BlockedLayout([2, 2], [4, THREADS_PER_WARP // 4], [1, 1], [0, 1], [2, 2], [2, 2], [1, 0]),
+    BlockedLayout([2, 2], [8, THREADS_PER_WARP // 8], [1, 1], [0, 1], [2, 2], [2, 2], [1, 0]),
+    BlockedLayout([1, 4], [1, THREADS_PER_WARP // 1], [1, 1], [0, 1], [1, 4], [1, 4], [1, 0]),
```
do we really need to test all those combinations?

ThomasRaoux reviewed (commented):
I don't think the changes are correct if we only support cases where the scan is within a CTA.

ThomasRaoux commented:
closing as there has been no activity for a while
